build/preprocessor.js

```diff
@@ -0,0 +1,98 @@
+'use strict';
+
+/* jshint node: true */
+/* global exports,
+          require */
+var utils = require('utils');
+
+const jsSuffix = /\.js$/;
+const htmlSuffix = /\.(html|htm)$/;
+
+function removeFiles(stagePath, list) {
+  list.forEach((fileName) => {
+    fileName.unshift(stagePath);
+    var file = utils.getFile.apply(this, fileName);
+    utils.log('PREPROCESSOR', 'remove file:', file.path);
+    if (utils.fileExists(file.path)) {
+      file.remove();
+    }
+  });
+}
+
+function processContent(flag, enable, content, type) {
+  var replaced;
+  if (enable) {
+    if (type === 'html') {
+      replaced = content.
+        replace('<!--IFDEF_' + flag, '', 'g').
+        replace('ENDIF_' + flag + '-->', '', 'g');
+    } else if (type === 'js') {
+      replaced = content.
+        replace('//IFDEF_' + flag, '', 'g').
+        replace('//ENDIF_' + flag, '', 'g');
+    }
+  } else {
+    var regexp;
+    if (type === 'html') {
+      regexp = new RegExp(
+        '<!--IFDEF_' + flag + '[^]*?ENDIF_' + flag + '-->', 'mg'
+      );
+    } else if (type === 'js') {
+      regexp = new RegExp(
+        '//IFDEF_' + flag + '[^]*?//ENDIF_' + flag, 'mg'
+      );
+    }
+    replaced = content.replace(regexp, '');
+  }
+
+  return replaced;
+}
+
+function processFiles(flag, enable, list, stagePath) {
+  list.forEach((fileName) => {
+    fileName.unshift(stagePath);
+    var file = utils.getFile.apply(this, fileName);
+    var fileContent = utils.getFileContent(file);
+    var type;
+    if (htmlSuffix.test(file.path)) {
+      type = 'html';
+    } else if (jsSuffix.test(file.path)) {
+      type = 'js';
+    }
+    var replacedContent = processContent(flag, enable, fileContent, type);
+    utils.writeContent(file, replacedContent);
+  });
+}
+
+/*
+  Preprocessor module provides a C/C++ style preprocessor mechanism to
+  customize HTML/Javascript source codes at build stage with a build flag.
+
+  Here is the file list example:
+  var fileList = {
+    process:[ // Process all IFDEF/ENDIF tag for each source code in the array.
+      ['index.html'],
```

Comment:
I prefer to specify these paths as absolute path instead of concating with stagePath inside here.